### ISSUE TITLE
variable weight LinearForm

### ISSUE THREAD

**Issue by M8kmyday** [mfem/mfem#3518]
I am solving a 3D electromagnetic problem with 2D ports on the boundary that have known field distributions but with unknown weights.  The goal is to solve for the 3D field distribution plus the unknown weights at the ports.  A typical boundary value problem with known boundary values leads to the Ax=b problem, for which you have the FormLinearSystem method.  With unknown weights, the problem becomes something like Ax=a1\*b1+a2\*b2+..._aN\*bN, where ai are the unknown weights and bj are known.  Setting one weight to 1, say a1 for discussion purposes, leads to Ax-a2\*b2-...-aN\*bN=a1\*b1.  In matrix form, this could be

```
[A  -B] [x]  = B [ak]
        [au]
```

where au are the unknown ai, ak is the known ai, and B is an array where column j equals bj.

To solve this problem, I'm trying to figure out how to use the FormLinearSystem method or its sub-methods to take maximum advantage of MFEM methods.  I don't see a way to do this, though.  Do you have any suggestions?

**Comment by mlstowell** [comment]
Hi, @M8kmyday,

This sounds interesting but I have not tried it. I would recommend using FormLinearSystem with all of the ports specified as essential BCs but set the BC values for the ports with unknown weights to zero. This should create a properly altered system matrix and a suitable right-hand-side vector. Next, the tricky part, grab the eliminated matrix using `ParBilinearForm::ParallelAssembleElim()` (if you're writing a serial app there should be a similarly named method). Multiply this "elim" matrix in front of your `B` matrix in the left hand side block system and this should complete the first row of the block system.

This would leave you with a rectangular system for which I assume you have a strategy in mind.

Best wishes,
Mark

**Comment by M8kmyday** [comment]
Thanks.

I can see why you called the ParallelAssemblyElim step "tricky".  I am only able to get a null HypreParMatrix pointer to be returned by the method.  I do not see an example or miniapp that uses ParallelAssemblyElim.  Do you happen to have a code snip that uses it?  I'll dig around in ParallelAssemblyElim and see if I can work out why mat_e isn't being set.

**Comment by mlstowell** [comment]
Unfortunately, I don't know of any code examples that use this. As I look more closely now I think I see why it does not work although I do not fully understand the implementation or purpose of `ParallelAssembleElim` (it clearly doesn't do what I assumed it did).

In serial `BilinearForm` has an internal matrix called `mat_e`. In parallel `ParBilinearForm::ParallelAssembleElim` combines the `mat_e` matrices from each processor into a single `HypreParMatrix`. Unfortunately, `mat_e` is not used in `ParBilinearForm::FormLinearSystem`, which uses `p_mat_e`, so it is empty. I believe what you want is `p_mat_e` but there doesn't seem to be any method to grab this private data.

As a hack you could add a method to `ParBilinearForm` which returns the pointer `p_mat_e`. This should allow you to experiment with your formulation. In the meantime the MFEM developers should try to figure out a more permanent solution.

**Comment by M8kmyday** [comment]
I derived a class and pulled out p_mat_e.  I wasn't able to get my setup working with FormLinearSystem, but that is on me.  In digging through FormLinearSystem and the functions it calls, I worked out how to build what I need doing my own elimination.  So I in effect built my own FormLinearSystem but it is customized to also solve for the unknown weights at the ports.  Thanks for the inputs and tips.
